Add icons related to linter messages in FileTreeNode

willdurand opened the pull request:
Fixes #329 
Fixes #330 

~~⚠️ depends on #400~~

---

This patch adds icons/colors to indicate when a file has linter messages. Directories containing such files are also marked with icons/colors.

Additional notes:

- only the most severe linter message type is shown
- HTML titles are used to give more information about the icons
- the "error" type is usually not used because errors block the submission of a version (admins can override this behavior, though)
- the first two screenshots show a yellow color for the warnings but it has been change to orange to be more readable
- I imagine that #115 could also be implemented with an icon (and some more style probably) **UPDATE:** see https://github.com/mozilla/addons-code-manager/pull/418/commits/04cac0a2ab0f23e38540787f7cdfc44a87a82a18

## Screenshot

With warning and notice messages shown. Because warnings are more severe than notices, the `chrome` folder has the "warning" icon:

![Screen Shot 2019-03-15 at 13 34 17](https://user-images.githubusercontent.com/217628/54436277-806bdc00-4732-11e9-8b97-1629299eddb7.png)

With error messages (please see the additional notes above):

<img width="1392" alt="Screen Shot 2019-03-19 at 11 47 50" src="https://user-images.githubusercontent.com/217628/54600382-edd28200-4a3c-11e9-8596-68f31558903f.png">

Updated color for the warnings:

![Screen Shot 2019-03-19 at 13 42 10](https://user-images.githubusercontent.com/217628/54607272-45c5b480-4a4e-11e9-868c-e390af7527cb.png)

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -178,4 +247,177 @@ describe(__filename, () => {
+      expect(type).toEqual(messages[1].type);
+    });
+
+    it('returns the most severe type given a map of messags and a full qualified file path', () => {
```
As far as I can tell there isn't a test to cover multiple messages at the same exact path

kumar303 commented on src/components/FileTreeNode/index.tsx:
```diff
@@ -3,31 +3,135 @@ import { TreefoldRenderProps } from 'react-treefold';
 
+    const hasLinterMessages =
+      linterMessages &&
+      Object.keys(linterMessages).some((path) => path.startsWith(node.id));
```
This doesn't seem to have test coverage. The tests still pass:

```diff
diff --git a/src/components/FileTreeNode/index.tsx b/src/components/FileTreeNode/index.tsx
index 092f273..8e5d41e 100644
--- a/src/components/FileTreeNode/index.tsx
+++ b/src/components/FileTreeNode/index.tsx
@@ -106,9 +106,7 @@ export class FileTreeNodeBase<TreeNodeType> extends React.Component<Props> {
       selectedPath,
     } = this.props;
 
-    const hasLinterMessages =
-      linterMessages &&
-      Object.keys(linterMessages).some((path) => path.startsWith(node.id));
+    const hasLinterMessages = !!linterMessages;
 
     let nodeIcons = null;
     let linterType = null;
```

kumar303 commented on src/components/FileTreeNode/index.tsx:
```diff
@@ -73,4 +179,19 @@ class FileTreeNodeBase<TreeNodeType> extends React.Component<PublicProps> {
+
+  const version = getVersionInfo(state.versions, versionId);
+  const selectedPath = version ? version.selectedPath : undefined;
+  const linterMessages = version
```
This patch introduces the same problem as https://github.com/mozilla/addons-code-manager/pull/400 but for `linterMessages`. The problem is that the component maps data from Redux but it doesn't know how to load that data when it doesn't exist. 

If this component can't load its own data then it has an implicit dependency on loader components such as [`Browse`](https://github.com/mozilla/addons-code-manager/blob/cd51cf5cea56138961e9eec7ff7d9ac193320369/src/pages/Browse/index.tsx#L86) or [`DiffView`](https://github.com/mozilla/addons-code-manager/blob/cd51cf5cea56138961e9eec7ff7d9ac193320369/src/components/DiffView/index.tsx#L104). The dependency is not explicit which makes it a fragile relationship. Let's say I move the loading logic *out* of `Browse` (which I plan to do); I would probably forget to check that this component still works.

In https://github.com/mozilla/addons-code-manager/pull/400 I suggested passing a direct public prop for what you need but if you did that here, you'd have to prop-drill `linterMessages` through `FileTree` which is not great. 

The best solution would be to implement [`<LinterProvider>`](https://github.com/mozilla/addons-code-manager/issues/414) because that's exactly what you need here. 

<details>
<summary>example patch</summary>

```diff
diff --git a/src/components/FileTreeNode/index.tsx b/src/components/FileTreeNode/index.tsx
index 092f273..dd976fc 100644
--- a/src/components/FileTreeNode/index.tsx
+++ b/src/components/FileTreeNode/index.tsx
@@ -92,14 +92,13 @@ const getIconForType = (type: string | null) => {
 };
 
 export class FileTreeNodeBase<TreeNodeType> extends React.Component<Props> {
-  render() {
+  renderWithMessages(linterMessages) {
     const {
       getToggleProps,
       hasChildNodes,
       isExpanded,
       isFolder,
       level,
-      linterMessages,
       node,
       onSelect,
       renderChildNodes,
@@ -177,6 +176,16 @@ export class FileTreeNodeBase<TreeNodeType> extends React.Component<Props> {
       </React.Fragment>
     );
   }
+
+  render() {
+    const { version } = this.props;
+
+    return (
+      <LinterProvider version={version}>
+        {(linterMessages) => this.renderWithMessages(linterMessages)}
+      </LinterProvider>
+    );
+  }
 }
 
 const mapStateToProps = (state: ApplicationState, ownProps: PublicProps) => {
```

</details>

It would introduce two linter message loaders on the same page but that's fine -- the first one to begin rendering will be the one that loads the data. The others will block on the `isLoading` flag.

The quickest solution would be to just use prop drilling and make a TODO comment to remove it once we have `<LinterProvider>`. I'm open to other ideas, though. Let me know if you have questions about why this is a problem.

kumar303 commented on stories/FileTreeNode.stories.tsx:
```diff
@@ -23,48 +28,47 @@ const getProps = ({
 };
 
+const render = ({ versionId = 123, store = configureStore(), ...props }) => {
+  return renderWithStoreAndRouter(
```
The `FileTree` stories also need to be wrapped in a store (depending on how you address previous comments). Right now they are broken for me with a red screen. Hmm. The tests should have broken for that; I'll file an issue.

kumar303 commented:
The bottom icon looks a little broken:

<img width="290" alt="Screenshot 2019-03-19 16 26 21" src="https://user-images.githubusercontent.com/55398/54642975-d490fc00-4a63-11e9-96ab-053fac20c792.png">

willdurand commented:
I believe I addressed most of the comments. I also made sure this view can be rendered on the Apple watch too 😁:

![Screen Shot 2019-03-20 at 13 18 46](https://user-images.githubusercontent.com/217628/54683843-0319e000-4b13-11e9-98d2-c2291a62de18.png)
![Screen Shot 2019-03-20 at 13 18 40](https://user-images.githubusercontent.com/217628/54683844-0319e000-4b13-11e9-9b58-f619941b5cc5.png)

Note that I did not add new stories yet, I'd like to do that in a follow-up if possible.

**Important:** prop-drilling did not work well for the "Compare" page because, for some reasons, I did not notice that `DiffView` was actually fetching/loading linter messages 😞 This does not look like a great idea and I am not sure connecting `FileTree` would be smart, unless we add `LinterProvider` everywhere, which does not sound like a great idea either.

I'd suggest we add a `linterMessagesByLine` prop in the `DiffView` component (like the `CodeView`, these two component APIs should be similar IMO) and we make `Compare` responsible for fetching/loading the linter messages as well as rendering the global messages (this code should be the same code as in `Browse`, so it means we could have a `GlobalLinterMessages` component). The `LinterProvider` could be used in both `Compare` and `Browse`.

willdurand commented:
rebased to reduce the diff size :)

kumar303 commented on src/components/FileTreeNode/styles.module.scss:
```diff
@@ -2,9 +2,32 @@
 }
 
+.nodeName {
+  word-break: break-all;
```
These styles don't look so nice on laptop-sized windows. How about this patch? It's a bit of an improvement.

<details>
<summary>patch</summary>

```diff
diff --git a/src/components/FileTreeNode/index.tsx b/src/components/FileTreeNode/index.tsx
index d02a372..7067676 100644
--- a/src/components/FileTreeNode/index.tsx
+++ b/src/components/FileTreeNode/index.tsx
@@ -134,23 +134,17 @@ class FileTreeNodeBase<TreeNodeType> extends React.Component<PublicProps> {
       <React.Fragment>
         <ListGroup.Item {...listGroupItemProps}>
           <span
-            className={styles.nodeName}
-            style={{ paddingLeft: `${level * 20}px` }}
+            className={styles.nodeItem}
+            style={{ paddingLeft: `${level * 12}px` }}
           >
             {isFolder ? (
-              <React.Fragment>
-                <FontAwesomeIcon icon={isExpanded ? 'folder-open' : 'folder'} />
-                &nbsp;{node.name}
-              </React.Fragment>
+              <FontAwesomeIcon icon={isExpanded ? 'folder-open' : 'folder'} />
             ) : (
-              <React.Fragment>
-                <FontAwesomeIcon icon="file" />
-                &nbsp;{node.name}
-              </React.Fragment>
+              <FontAwesomeIcon icon="file" />
             )}
+            <span className={styles.nodeName}>{node.name}</span>
+            {nodeIcons}
           </span>
-
-          {nodeIcons}
         </ListGroup.Item>
 
         {isExpanded &&
diff --git a/src/components/FileTreeNode/styles.module.scss b/src/components/FileTreeNode/styles.module.scss
index 54d97e8..e0dd03b 100644
--- a/src/components/FileTreeNode/styles.module.scss
+++ b/src/components/FileTreeNode/styles.module.scss
@@ -9,8 +9,29 @@
   padding: 0.5rem 1.25rem;
 }
 
+.nodeItem {
+  align-items: flex-start;
+  display: flex;
+  line-height: 1;
+  width: 100%;
+}
+
+.nodeName,
+.nodeIcons {
+  margin-left: 0.5rem;
+  width: 100%;
+}
+
 .nodeName {
-  word-break: break-all;
+  overflow: hidden;
+  padding-bottom: 0.2rem;
+  text-overflow: ellipsis;
+  white-space: wrap;
+}
+
+.nodeIcons {
+  display: flex;
+  justify-content: flex-end;
 }
 
 .selected {
@@ -28,6 +49,3 @@
     color: $orange;
   }
 }
-
-.nodeIcons {
-}
```

</details>

<img width="568" alt="Screenshot 2019-03-20 12 57 59" src="https://user-images.githubusercontent.com/55398/54707850-0e1f4100-4b10-11e9-8852-7989d462458c.png">
<img width="965" alt="Screenshot 2019-03-20 12 57 44" src="https://user-images.githubusercontent.com/55398/54707852-0f506e00-4b10-11e9-9b61-7adc10a0a22d.png">


I filed https://github.com/mozilla/addons-code-manager/issues/431 to make a storybook out of this.

willdurand replied:
oh yes, I love it! Thanks ❤️

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -45,17 +50,42 @@ describe(__filename, () => {
+
+  const renderWithLinterMessage = ({
+    version = createInternalVersion(fakeVersion),
+    message = fakeExternalLinterMessage,
```
This seems like it would be better as an array but I guess if you don't need it then it's fine.

willdurand replied:
yep, I'll do it in the next PR

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -216,4 +246,200 @@ describe(__filename, () => {
+    expect(root.find(className)).toHaveLength(1);
+
+    const nodeIcons = root.find(`.${styles.nodeIcons}`);
+    expect(nodeIcons).toHaveLength(1);
```
You could use `.toHaveClassName()`

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -216,4 +246,200 @@ describe(__filename, () => {
+        },
+      });
+
+      expect(root.find(className)).toHaveLength(1);
```
`.toHaveClassName()` would give a better message here

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -216,4 +246,200 @@ describe(__filename, () => {
+    (type, className, icon) => {
+      const message = {
+        ...fakeExternalLinterMessage,
+        file: 'manifest.json',
```
Could you make this more realistic for the test? Was this path supposed to be a directory?

willdurand replied:
yep, thanks

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -216,4 +246,200 @@ describe(__filename, () => {
+      },
+    });
+
+    expect(root.find(`.${styles.hasLinterMessages}`)).toHaveLength(0);
```
I won't comment on them all but `.toHaveClassName()` would produce better messages in all of the assertions I'm seeing.

willdurand replied:
`toHaveClassName()` and `.find()` are different. One is looking for a class name on the current wrapper, the other is used to make sure a component identified by its class name exits.

kumar303 commented on src/components/FileTreeNode/index.spec.tsx:
```diff
@@ -216,4 +246,200 @@ describe(__filename, () => {
+          file: `${path}/file-2.js`,
+          type: 'error',
+        },
+      ] as ExternalLinterMessage[];
```
This is not a change request but since this has come up a lot, maybe we should switch to string enums for `type`.

willdurand replied:
it's a bummer that it does not understand this notation...

kumar303 commented on src/pages/Compare/index.tsx:
```diff
@@ -31,7 +31,7 @@ type PropsFromRouter = {
 type PropsFromState = {
   addonId: number;
-  compareInfo: CompareInfo | null | void;
+  compareInfo: CompareInfo | null | undefined;
```
I thought `void` was preferred, no? If it's trying to match another type, maybe change that one?

willdurand replied:
ahm, this is not needed actually
